### PR TITLE
LUCENE-9507 Custom order for leaves in IndexReader and IndexWriter

### PR DESCRIPTION
**Pull request by mayya-sharipova**
1. Add an option to supply a custom leaf sorter for IndexWriter.
A DirectoryReader opened from this IndexWriter will have its leaf
readers sorted with the provided leaf sorter. This is useful for
indices on which it is expected to run many queries with particular
sort criteria (e.g. for time-based indices this is usually a
descending sort on timestamp). Providing leafSorter allows
to speed up early termination for this particular type of
sort queries.

2. Add an option to supply a custom sub-readers sorter for
BaseCompositeReader. In this case sub-readers will be sorted 
according to the the provided leafSorter.

3. Add an option to supply a custom leaf sorter for
StandardDirectoryReader. The leaf readers of this
StandardDirectoryReader will be sorted according to
the the provided leaf sorter.

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/index/IndexWriterConfig.java [hunk 1]
```diff
@@ -53,7 +54,7 @@
 public final class IndexWriterConfig extends LiveIndexWriterConfig {
 
   /** Specifies the open mode for {@link IndexWriter}. */
-  public static enum OpenMode {
+  public enum OpenMode {
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/IndexWriterConfig.java:57]: Silly java, why doesn't `javac` warn/tell us that `static` is pointless for `enum`?  Hmm, it is pointless, right?

Wow, we have tons of `static enum` in our codebase!  I'll open a separate PR, this is crazy.

File: lucene/core/src/java/org/apache/lucene/index/LiveIndexWriterConfig.java [hunk 2]
```diff
@@ -91,6 +92,9 @@
   /** The sort order to use to write merged segments. */
   protected Sort indexSort = null;
 
+  /** The comparator for sorting leaf readers. */
+  protected Comparator<LeafReader> leafSorter = null;
```
Comment (mikemccand) [inline lucene/core/src/java/org/apache/lucene/index/LiveIndexWriterConfig.java:96]: No need for `= null` -- it is default already for java.

**Review by mikemccand — APPROVED** [review]
This looks like a nice feature!  Thank you @mayya-sharipova!

I left small comments that shouldn't block pushing this cool new feature.

One warning: our reader constructor/factory API is crazy complex and it's easy with an addition like this to accidentally break backwards compatibility in the public API.  From what I saw you succeeded in preserving backwards compatibility, except for internal / protected constructors, and so we could backport this to 8.x too, but maybe double check that ;)

File: lucene/core/src/java/org/apache/lucene/index/DirectoryReader.java [hunk 3]
```diff
@@ -56,7 +57,24 @@
...
+  /**
+   * Returns a IndexReader for the the index in the given Directory
+   *
+   * @param directory the index directory
+   * @param leafSorter a comparator for sorting leaf readers. Providing leafSorter is useful for
+   *     indices on which it is expected to run many queries with particular sort criteria (e.g. for
+   *     time-based indices this is usually a descending sort on timestamp). In this case {@code
+   *     leafSorter} should sort leaves according to this sort criteria. Providing leafSorter allows
+   *     to speed up this particular type of sort queries by early terminating while iterating
+   *     through segments and segments' documents.
```
[the remark this answers is not in the document]
Comment (jimczi) [reply lucene/core/src/java/org/apache/lucene/index/DirectoryReader.java:72]: It seems to work well in practice with the tiered merge policy and natural merging. The natural order of insertion seems to be preserved nicely in the levels. Things get complicated when sorting on descending order though. The big segments tend to have the data in ascending order. So in this scenario, it's much slower to retrieve the latest documents if the index was force merged to 1 segment or few big ones. In Elasticsearch we shuffle the order of the segments on force merge to alleviate this bias. 
In this PR we opted for a single sort per index so I like the idea of having a merge policy that is optimized for this use case. I think it's a good compromise for cases where index sorting is too costly at the document level.

File: lucene/core/src/java/org/apache/lucene/index/BaseCompositeReader.java [hunk 4]
```diff
@@ -231,4 +239,13 @@ protected final int readerBase(int readerIndex) {
...
     return subReadersList;
   }
+
+  /**
+   * Returns a comparator for sorting sub readers.
+   *
+   * @return a comparator for sorting sub readers, or {@code null} if a comparator was not set.
+   */
+  protected Comparator<R> getSubReadersSorter() {
+    return subReadersSorter;
```
[the remark this answers is not in the document]
Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/index/BaseCompositeReader.java]: > .. if we wrap this reader in a FilterDirectoryReader, that wrapped reader will also sort its leaves the same way?

Thank you , that's a very good point, you are right, indeed the wrapped reader will already sort its leaves the same way.  Fixed by removing `sorter` in `FilterDirectoryReader` in 34da11d80d30c1d4a418066edfadce0dde3379ae

> .. why does the reader need to record its sorter? I guess maybe callers might want to "chain" the sorter -- have this reader remember what sorter was used, and on refresh, pull that sorter out and use it again for the next reader, ok.

Right, indeed.  In `StandardDirectoryReader::doOpenIfChanged`, we reuse sorter.

File: lucene/core/src/java/org/apache/lucene/index/LiveIndexWriterConfig.java (hunk 2 above)
Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/index/LiveIndexWriterConfig.java]: Fixed in 34da11d80d30c1d4a418066edfadce0dde3379ae

**Comment by mayya-sharipova** [comment]
@mikemccand Thanks a lot for your feedback, addressed in 34da11d80d30c1d4a418066edfadce0dde3379ae

File: lucene/core/src/java/org/apache/lucene/index/BaseCompositeReader.java [hunk 5]
```diff
@@ -46,6 +47,7 @@
  */
 public abstract class BaseCompositeReader<R extends IndexReader> extends CompositeReader {
   private final R[] subReaders;
+  final Comparator<R> subReadersSorter;
```
Comment (uschindler) [inline lucene/core/src/java/org/apache/lucene/index/BaseCompositeReader.java:50]: Why is this package protected? This is a public class, so if it is needed by subclasses, it needs to be protected. If it is not needed by subclasses, make it private.

**Review by uschindler — CHANGES_REQUESTED** [review]

File: lucene/core/src/java/org/apache/lucene/index/BaseCompositeReader.java (hunk 5 above)
Comment (mayya-sharipova) [reply lucene/core/src/java/org/apache/lucene/index/BaseCompositeReader.java]: @uschindler Thank you for checking. That's a great observation. As the field is used by subclasses, I made it `protected` in d9159c7e13492d22acf0003e262074479314cf4b.

Comment (uschindler) [reply lucene/core/src/java/org/apache/lucene/index/BaseCompositeReader.java]: Thanks, much better now!

**Pull request merged**